# Function Assingment 1

1. write a Python function to find the maximum of three numbers.

In [1]:
def maximum(a,b,c):
    if a > b and a > c:
        greater = a
    elif b > a and b > c:
        greater = b
    elif c > a and c > b:
        greater = c
    else:
       greater = 'Numbers are equal'
    return greater

In [3]:
print(maximum(2,1,1))

2


2. Write a Python function to sum all the numbers in a list.

    Sample List : (8, 2, 3, 0, 7)

    Expected Output : 20

In [6]:
def add(a):
    addition = 0
    for i in a:
        addition += i
    return addition

In [8]:
lst = [8,2,3,0,7]

print(add(lst))

20


3. Write a Python function to multiply all the numbers in a list.

    Sample List : (8, 2, 3, -1, 7)

    Expected Output : -336

In [13]:
def multiply(a):
    multi = 1
    for i in a:
        multi = multi * i
    return multi

In [15]:
lst1 = [8, 2, 3, -1, 7] 

print(multiply(lst1))

-336


4. Write a Python program to reverse a string.
    
    Sample String : "1234abcd"
   
    Expected Output : "dcba4321"

In [18]:
def reverseStr(a):
    new = a[len(a)::-1]
    return new

In [20]:
print(reverseStr('1234abcd'))

dcba4321


5. Write a Python function to calculate the factorial of a number (a non-negative
integer). The function accepts the number as an argument.

In [23]:
def factorial(a):
    fact = 1
    for i in range(1,a+1):
        fact *= i
    return fact

In [25]:
print(factorial(6))

720


6. Write a Python function to check whether a number falls within a given range.

In [28]:
def is_in_range(number, start, end):
    return start <= number <= end

In [30]:
number = 16
range_start = 5
range_end = 15

if is_in_range(number, range_start, range_end):
    print(f"{number} is within the range [{range_start}, {range_end}].")
else:
    print(f"{number} is NOT within the range [{range_start}, {range_end}].")


16 is NOT within the range [5, 15].


7. Write a Python function that accepts a string and counts the number of upper
and lower case letters.

    Sample String : 'The quick Brow Fox'

    Expected Output:

    - No. of Upper case characters : 3
    - No. of Lower case Characters : 12

In [22]:
def count_low_up(a):
    newstr = a.replace(' ','')
    up = 0
    lp = 0
    for i in newstr:
        if i == i.lower():
            lp += 1
        else:
            up += 1
    return f'''
Lower Count = {lp}
Upper Count = {up}
'''

In [24]:
print(count_low_up('My Name is Tayyab'))


Lower Count = 11
Upper Count = 3



8. Write a Python function that takes a list and returns a new list with distinct

    elements from the first list.

    Sample List : [1,2,3,3,3,3,4,5]

    Unique List : [1, 2, 3, 4, 5]

In [38]:
def distinct(a):
    return list(set(a))

In [40]:
Sample = [1,2,3,3,3,3,4,5]
distinct(Sample)

[1, 2, 3, 4, 5]

9. Write a Python function that takes a number as a parameter and checks
whether the number is prime or not.

    Note : A prime number (or a prime) is a natural number greater than 1 and that has no positive divisors other than 1 and itself.

In [70]:
def check_prime(a):
    if a <= 1:
        return "Not a Prime Number"
    if a == 2:
        return "Prime Number"
    if a % 2 == 0:
        return "Not a Prime Number"
    for i in range(3, int(a**0.5) + 1, 2):
        if a % i == 0:
            return "Not a Prime Number"
    return "Prime Number"

In [72]:
print(check_prime(129))

Not a Prime Number


10. Write a Python program to print the even numbers from a given list.

    Sample List : [1, 2, 3, 4, 5, 6, 7, 8, 9]

    Expected Result : [2, 4, 6, 8]

In [144]:
def even_list(a):
    new = []
    for i in a:
        if i%2 == 0:
            new.append(i)
    return new

In [141]:
sampleList = [1, 2, 3, 4, 5, 6, 7, 8, 9]
even_list(sampleList)

[2, 4, 6, 8]

11. Write a Python function to check whether a number is "Perfect" or not.
According to Wikipedia : In number theory, a perfect number is a positive integer
that is equal to the sum of its proper positive divisors, that is, the sum of its
positive divisors excluding the number itself (also known as its aliquot sum).
Equivalently, a perfect number is a number that is half the sum of all of its
positive divisors (including itself).

    Example : The first perfect number is 6, because 1, 2, and 3 are its proper positive divisors, and 1 + 2 + 3 = 6. Equivalently, the number 6 is equal to half the sum of all its positive divisors: ( 1 + 2 + 3 + 6 ) / 2 = 6. The next perfect number is 28 = 1 + 2 + 4 + 7 + 14. This is followed by the perfect numbers 496 and 8128.

In [206]:
def perfect(a):
    s = 0
    for i in range(1,a+1):
        if a%i == 0:
            s += i
    if s/2 == a:
        print("Perfect")
    else:
        print("Not a perfect")

In [214]:
perfect(10)
perfect(496)
perfect(30)
perfect(8128)

Not a perfect
Perfect
Not a perfect
Perfect


12. Write a Python function that checks whether a passed string is a palindrome
or not.

    Note: A palindrome is a word, phrase, or sequence that reads the same backward as forward, e.g., madam or nurses run.

In [217]:
def palindrome(a):
    new = a[len(a)::-1]
    if new == a:
        print('Palindrome')
    else:
        print('Not a palindrome')

In [225]:
palindrome('madam')
palindrome('nurses')
palindrome('radar')
palindrome('run')

Palindrome
Not a palindrome
Palindrome
Not a palindrome


# Function Assingment 2

1. You need to write a function in python that return all such numbers which are
divisible by 7 but are not a multiple of 5, between 2000 and 3200 (both included).
Return the obtained numbers in a comma-separated sequence on a single line.

In [253]:
def task1(start, end):
    result = []
    for i in range(start, end + 1):
        if i % 7 == 0 and i % 5 != 0:
            result.append(str(i))
    return ','.join(result)

In [255]:
print(task1(2000,3200))

2002,2009,2016,2023,2037,2044,2051,2058,2072,2079,2086,2093,2107,2114,2121,2128,2142,2149,2156,2163,2177,2184,2191,2198,2212,2219,2226,2233,2247,2254,2261,2268,2282,2289,2296,2303,2317,2324,2331,2338,2352,2359,2366,2373,2387,2394,2401,2408,2422,2429,2436,2443,2457,2464,2471,2478,2492,2499,2506,2513,2527,2534,2541,2548,2562,2569,2576,2583,2597,2604,2611,2618,2632,2639,2646,2653,2667,2674,2681,2688,2702,2709,2716,2723,2737,2744,2751,2758,2772,2779,2786,2793,2807,2814,2821,2828,2842,2849,2856,2863,2877,2884,2891,2898,2912,2919,2926,2933,2947,2954,2961,2968,2982,2989,2996,3003,3017,3024,3031,3038,3052,3059,3066,3073,3087,3094,3101,3108,3122,3129,3136,3143,3157,3164,3171,3178,3192,3199


2. You need code a function that calculates and returns the value according to the given formula:

P = Square root of [(2 * A * B)/C]

Following are the fixed values of A and B:

A is 50. B is 30.

The values of the literal C should be taken as console input to your program in a
comma-separated sequence.

Example
Let us assume the following comma separated input sequence is given to the program:

100,150,180

The output of the program should be:

18,22,24

In [319]:
import math

def task2():
    A = 50
    B = 30
    inp = input("Enter comma-separated values for C: ")
    C = [int(i) for i in inp.split(",")]
    results = []
    for j in C:
        P = math.sqrt((2 * A * B) / j)
        results.append(str(int(P)))
    return ",".join(results)

In [321]:
print(task2())

Enter comma-separated values for C:  100,150,180


5,4,4


3. You need to write a function that takes a comma separated sequence of words as input and prints the words in a comma-separated sequence after sorting them alphabetically.

Suppose the following input is supplied to the program:

without,hello,bag,world

Then, the output should be:

bag,hello,without,world

In [354]:
def task3():
    inp = input("Enter words: ")
    lst = inp.split(",")
    n = len(lst)
    for i in range(n):
        for j in range(0, n-i-1):
            if lst[j] > lst[j+1]:
                lst[j], lst[j+1] = lst[j+1], lst[j]
    print(",".join(lst))

In [356]:
task3()

Enter words:  without,hello,bag,world


bag,hello,without,world


4. You need to write a program that takes sequence of lines as input and prints the lines after making all characters in the sentence capitalized.

Suppose the following input is supplied to the program:

Hello world

Practice makes perfect

Then, the output should be:

HELLO WORLD

PRACTICE MAKES PERFECT

In [362]:
def task4():
    print("Enter the lines (doublepress Enter to stop):")
    lines = []
    
    while True:
        line = input()
        if line == "":
            break
        lines.append(line)

    for line in lines:
        print(line.upper())

In [364]:
task4()

Enter the lines (doublepress Enter to stop):


 Hello, 
 my name is tayyab 
 today is friday
 


HELLO, 
MY NAME IS TAYYAB 
TODAY IS FRIDAY


5. You need to write a function that counts the number of vowels in a given sentence as input from console.

Suppose the following input is supplied to the program:

Hello world

Practice makes perfect

Then, the output should be:

    - a appeared 2 times
    - e appeared 5 times
    - i appeared 1 time
    - o appeared 2 times
    - u appeared 0 time

In [387]:
def task5():
    sentence = input("Enter a sentence: ")
    vowels = 'aeiou'
    vowel_dis = {vowel: 0 for vowel in vowels}
    
    for i in sentence.lower():
        if i in vowels:
            vowel_dis[i] += 1
    for vowel in vowels:
        if vowel_dis[vowel] == 1:
            print(f"{vowel} appeared {vowel_dis[vowel]} time")
        else:
            print(f"{vowel} appeared {vowel_dis[vowel]} times")

In [389]:
task5()

Enter a sentence:  Hello world  Practice makes perfect


a appeared 2 times
e appeared 5 times
i appeared 1 time
o appeared 2 times
u appeared 0 times


6. You need write a function that traces and makes a list of all such numbers
from1000 to 3000 in which all the digits are even numbers.

In [392]:
def task6(start,end):
    lst = []
    for i in range(start,end+1):
        if i%2 == 0:
            lst.append(i)
    return lst

In [396]:
print(task6(1000,3000))

[1000, 1002, 1004, 1006, 1008, 1010, 1012, 1014, 1016, 1018, 1020, 1022, 1024, 1026, 1028, 1030, 1032, 1034, 1036, 1038, 1040, 1042, 1044, 1046, 1048, 1050, 1052, 1054, 1056, 1058, 1060, 1062, 1064, 1066, 1068, 1070, 1072, 1074, 1076, 1078, 1080, 1082, 1084, 1086, 1088, 1090, 1092, 1094, 1096, 1098, 1100, 1102, 1104, 1106, 1108, 1110, 1112, 1114, 1116, 1118, 1120, 1122, 1124, 1126, 1128, 1130, 1132, 1134, 1136, 1138, 1140, 1142, 1144, 1146, 1148, 1150, 1152, 1154, 1156, 1158, 1160, 1162, 1164, 1166, 1168, 1170, 1172, 1174, 1176, 1178, 1180, 1182, 1184, 1186, 1188, 1190, 1192, 1194, 1196, 1198, 1200, 1202, 1204, 1206, 1208, 1210, 1212, 1214, 1216, 1218, 1220, 1222, 1224, 1226, 1228, 1230, 1232, 1234, 1236, 1238, 1240, 1242, 1244, 1246, 1248, 1250, 1252, 1254, 1256, 1258, 1260, 1262, 1264, 1266, 1268, 1270, 1272, 1274, 1276, 1278, 1280, 1282, 1284, 1286, 1288, 1290, 1292, 1294, 1296, 1298, 1300, 1302, 1304, 1306, 1308, 1310, 1312, 1314, 1316, 1318, 1320, 1322, 1324, 1326, 1328, 1330, 133

7. You need to write a code which accepts a sequence of comma separated 4 digit binary numbers as its input and then check whether they are divisible by 5 or not.

The numbers that are divisible by 5 are to be printed in a comma separated sequence.

Example:

0100,0011,1010,1001

Then the output should be:

1010

In [399]:
def task7():
    number = input("Enter comma-separated 4-digit binary numbers: ")
    
    binary_list = number.split(",")
    divisible_by_5 = []
    for binary in binary_list:
        decimal = int(binary, 2)
        if decimal % 5 == 0:
            divisible_by_5.append(binary)
    print(",".join(divisible_by_5))

In [403]:
task7()

Enter comma-separated 4-digit binary numbers:  0100,0011,1010,1001


1010


8. Write a program that accepts a sentence and calculate the number of letters and digits.

Suppose the following input is supplied to the program:

hello world! 123

Then, the output should be:

LETTERS 10

DIGITS 3

In [407]:
def task8():
    sentence = input("Enter a sentence: ")
    
    letters = 0
    digits = 0
    for char in sentence:
        if char.isalpha():
            letters += 1
        elif char.isdigit():
            digits += 1
    
    print(f"LETTERS {letters}")
    print(f"DIGITS {digits}")

In [409]:
task8()

Enter a sentence:  hello world! 123


LETTERS 10
DIGITS 3
